node/smeshing_service_test.go

```diff
@@ -0,0 +1,240 @@
+		defer cancel()
+
+		errCh := make(chan error)
+		go func() {
+			errCh <- service.Start(ctx)
+		}()
+		select {
+		case <-service.started:
+		case <-ctx.Done():
+			t.Fatal("not started on time")
+		}
+
+		conn, err := grpc.NewClient(
+			service.grpcPublicServer.BoundAddress,
+			grpc.WithTransportCredentials(insecure.NewCredentials()),
+		)
+		require.NoError(t, err)
+		t.Cleanup(func() { require.NoError(t, conn.Close()) })
+
+		client := pb.NewDebugServiceClient(conn)
+		resp, err := client.ChangeLogLevel(ctx, &pb.ChangeLogLevelRequest{
+			Module: "poet",
+			Level:  "DEBUG",
+		})
+		require.NoError(t, err)
+		require.NotNil(t, resp)
+	})
+}
+
+func TestSmeshingService_StartSmeshing(t *testing.T) {
+	t.Run("starts smeshing with valid coinbase", func(t *testing.T) {
+		cfg := getSmeshingServiceTestConfig(t)
+		service, err := NewSmeshingService(cfg, zaptest.NewLogger(t))
+		require.NoError(t, err)
+
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+		defer cancel()
+
+		err = service.startServices(ctx)
+		require.NoError(t, err)
+	})
+
+	t.Run("fails with invalid coinbase", func(t *testing.T) {
+		cfg := getSmeshingServiceTestConfig(t)
+		cfg.SMESHING.CoinbaseAccount = "invalid-address"
+
+		service, err := NewSmeshingService(cfg, zaptest.NewLogger(t))
+		require.NoError(t, err)
+
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+		defer cancel()
+
+		err = service.startServices(ctx)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "parse CoinbaseAccount address")
+	})
+}
+
+func TestSmeshingService_GRPCServices(t *testing.T) {
+	t.Run("initializes all grpc services", func(t *testing.T) {
+		cfg := getSmeshingServiceTestConfig(t)
+		service, err := NewSmeshingService(cfg, zaptest.NewLogger(t))
+		require.NoError(t, err)
+
+		services := []grpcserver.Service{
+			grpcserver.Debug,
+			grpcserver.Smesher,
+			grpcserver.Post,
+			grpcserver.PostInfo,
+		}
+
+		for _, svc := range services {
+			api, err := service.grpcService(svc, zaptest.NewLogger(t))
+			require.NoError(t, err)
+			require.NotNil(t, api)
+		}
+	})
+
+	t.Run("fails with unknown service", func(t *testing.T) {
+		cfg := getSmeshingServiceTestConfig(t)
+		service, err := NewSmeshingService(cfg, zaptest.NewLogger(t))
+		require.NoError(t, err)
+
+		api, err := service.grpcService("unknown", zaptest.NewLogger(t))
+		require.Error(t, err)
+		require.Nil(t, api)
+	})
+}
+
+func TestSmeshingService_StopServices(t *testing.T) {
+	t.Run("stops all services", func(t *testing.T) {
+		cfg := getSmeshingServiceTestConfig(t)
+		service, err := NewSmeshingService(cfg, zaptest.NewLogger(t))
+		require.NoError(t, err)
+
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+
+		err = service.start(ctx)
+		require.NoError(t, err)
+
+		stopCtx, stopCancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer stopCancel()
+
+		service.stopServices(stopCtx)
```

Comment:
Same issue here: if `start` is not blocking, then how to properly stop the service?

- `stopServices`? but then what is the `ctx` needed for?
- canceling the context? but then what is `stopServices` needed for?
- both? What if only one of the two is called?